src/GramatykiGrafoweAGH/__init__.py

```diff
@@ -1,2 +1,177 @@
-def hello_world():
-    return "Hello world!"
+from __future__ import annotations
+
+from collections import defaultdict
+from typing import Iterable
+from dataclasses import dataclass, field
+from itertools import count
+from typing import List, Tuple, Callable
+
+import networkx as nx
+
+from GramatykiGrafoweAGH.exceptions import NodeNotFoundError, CannotApplyProductionError
+
+
+@dataclass(eq=True, frozen=True)
+class Node:
+    label: str
+    x: float
+    y: float
+    level: int
+    id: int = field(default_factory=count().__next__, init=False)
+
+
+class NodePositions:
+    def __init__(self):
+        self._dict = defaultdict(list)
+
+    @staticmethod
+    def key(node: Node) -> Tuple[int, float, float]:
+        return node.level, node.x, node.y
+
+    def add_node(self, node: Node) -> None:
+        self._dict[self.key(node)].append(node)
+
+    def remove_node(self, node: Node) -> None:
+        self._dict[self.key(node)].remove(node)
+
+    def get_duplicates_of(self, node: Node) -> Iterable[Node]:
+        group = self._dict[node.level, node.x, node.y]
+        return [n for n in group if n is not node and n.label == node.label]
+
+    def get_duplicates_with_label(self, label: str) -> Iterable[List[Node]]:
+        for group in self._dict.values():
+            nodes = [n for n in group if n.label == label]
+            if len(nodes) >= 2:
+                yield nodes
+
+    def count_duplicates(self) -> int:
+        return sum(1 for group in self._dict.values() if len(group) >= 2)
+
+
+class Graph:
+    def __init__(self):
+        self._G = nx.Graph()
+        self._node_positions = NodePositions()
+
+    def __contains__(self, node: Node) -> bool:
+        return self.has_node(node)
+
+    @property
+    def nodes(self):
```

Comment:
no type

Reply:
Feel free to fill it.

Reply:
No

Reply:
Fixed in ea36104b7e70f0614cd130c9337693f412f3753c